runtime/runtime/src/genesis.rs

```diff
@@ -71,210 +69,306 @@ impl<'a> StorageComputer<'a> {
     }
 
     /// Adds multiple StateRecords to the users' storage info.
-    pub fn process_records(&mut self, records: &[StateRecord]) {
+    pub(crate) fn process_records(&mut self, records: &[StateRecord]) {
         for record in records {
             self.process_record(record);
         }
     }
 
     /// Returns the current storage use for each user.
-    pub fn finalize(self) -> HashMap<AccountId, u64> {
+    pub(crate) fn finalize(self) -> HashMap<AccountId, u64> {
         self.result
     }
 }
 
-pub struct GenesisStateApplier {}
+const MAX_OUTSTANDING_WRITES: usize = 0x100_000;
+type State = (usize, StateRoot, TrieUpdate);
 
-impl GenesisStateApplier {
-    fn commit(
-        mut state_update: TrieUpdate,
-        current_state_root: &mut StateRoot,
-        tries: &mut ShardTries,
-        shard_uid: ShardUId,
-    ) {
+pub(crate) struct AutoFlushingTrieUpdate<'a> {
+    active_writers: &'a atomic::AtomicUsize,
+    tries: &'a ShardTries,
+    shard_uid: ShardUId,
+    state: Mutex<State>,
+}
+
+impl<'a> AutoFlushingTrieUpdate<'a> {
+    pub(crate) fn new(
+        active_writers: &'a atomic::AtomicUsize,
+        state_root: StateRoot,
+        tries: &'a ShardTries,
+        uid: ShardUId,
+    ) -> Self {
+        let state = Mutex::new((0, state_root, tries.new_trie_update(uid, state_root)));
+        Self { active_writers, state, tries, shard_uid: uid }
+    }
+
+    fn lock(&self) -> std::sync::MutexGuard<State> {
+        self.state.lock().unwrap_or_else(|g| g.into_inner())
```

Comment:
Mutex poisoning is one of those things whose utility is a subject of heated debate in the Rust community. See https://blog.rust-lang.org/2020/12/11/lock-poisoning-survey.html for example.

For nearcore specifically, `--panic=abort`, so poisoned mutexes are an impossible state in the first place. But otherwise the choice is quite arbitrary on my part here.

Reply:
I see, thanks for the explanation.